src/Gravity.hpp

```diff
@@ -0,0 +1,494 @@
+///
+
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <memory>
+
+#include "AMReX_Array.H"
+#include "AMReX_BCRec.H"
+#include "AMReX_BC_TYPES.H"
+#include "AMReX_BoxArray.H"
+#include "AMReX_Config.H"
+#include "AMReX_DistributionMapping.H"
+#include "AMReX_Geometry.H"
+#include "AMReX_GpuContainers.H"
+#include "AMReX_IntVect.H"
+#include "AMReX_MFIter.H"
+#include "AMReX_MultiFabUtil.H"
+#include "AMReX_PhysBCFunct.H"
+#include "AMReX_REAL.H"
+#include <AMReX_FillPatchUtil.H>
+#include <AMReX_MLLinOp.H>
+#include <AMReX_MLMG.H>
+#include <AMReX_MLPoisson.H>
+#include <AMReX_ParmParse.H>
+
+#include "simulation.hpp"
+
+// #define GRAVITY_DEBUG
+
+// This vector can be accessed on the GPU.
+using RealVector = amrex::Gpu::ManagedVector<amrex::Real>;
+using Real = amrex::Real;
+using Box = amrex::Box;
+using BoxArray = amrex::BoxArray;
+using MultiFab = amrex::MultiFab;
+using BCRec = amrex::BCRec;
+using ParmParse = amrex::ParmParse;
+using IntVect = amrex::IntVect;
+using Geometry = amrex::Geometry;
+using DistributionMapping = amrex::DistributionMapping;
+using MFIter = amrex::MFIter;
+template <typename T> using Array4 = amrex::Array4<T>;
+template <typename T, int N> using GpuArray = amrex::GpuArray<T, N>;
+template <typename T> using Vector = amrex::Vector<T>;
+
+///
+/// Gravity solve parameters
+///
+
+namespace gravity
+{
+enum class GravityMode {
+	Poisson,
+	Constant,
+};
+
+const GravityMode gravity_type = GravityMode::Poisson;
+const amrex::Real const_grav = 0.0;
+const int lnum = 16;
+const int max_solve_level = 10;
+
+const int verbose = 0;
+const int no_sync = 1;
+const int do_composite_phi_correction = 1;
+
+// multigrid solve parameters (all boolean)
+const int mlmg_agglomeration = 1;
+const int mlmg_consolidation = 1;
+const int mlmg_max_fmg_iter = 0;
+} // namespace gravity
+
+///
+/// Multipole gravity data
+///
+namespace multipole
+{
+const int lnum_max = 30;
+
+constexpr amrex::Real volumeFactor = 1.0;
+constexpr amrex::Real parityFactor = 1.0;
+
+amrex::Array1D<bool, 0, 2> constexpr doSymmetricAddLo = {false};
+amrex::Array1D<bool, 0, 2> constexpr doSymmetricAddHi = {false};
+bool constexpr doSymmetricAdd = false;
+
+amrex::Array1D<bool, 0, 2> constexpr doReflectionLo = {false};
+amrex::Array1D<bool, 0, 2> constexpr doReflectionHi = {false};
+
+extern AMREX_GPU_MANAGED amrex::Real rmax;
+
+extern AMREX_GPU_MANAGED amrex::Array2D<amrex::Real, 0, lnum_max, 0, lnum_max> factArray;
+extern AMREX_GPU_MANAGED amrex::Array1D<amrex::Real, 0, lnum_max> parity_q0;
+extern AMREX_GPU_MANAGED amrex::Array2D<amrex::Real, 0, lnum_max, 0, lnum_max> parity_qC_qS;
+} // namespace multipole
+
+///
+/// @class Gravity
+/// @brief
+///
+template <typename T> class Gravity
+{
+
+      public:
+	///
+	/// Constructor
+	///
+	/// @param _Density         index of density component
+	///
+	Gravity(AMRSimulation<T> *_sim, amrex::BCRec &phys_bc, amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> &_coordCenter, int Density_);
```

Comment:
You can add an additional argument for the cell-centered boundary conditions here.

They are not the same as `phys_bc`, since that is technically the boundary conditions for the Poisson equation. However, we don't really need `phys_bc`, since we only support vacuum boundary conditions for the Poisson equation.